### PR TITLE
Add remote address to failed/succeeded Bot/Instance joins

### LINKED ISSUE #41421: Include source info for node/bot join fails
### Expected behavior:
When viewing audit logs for instance join failures, the logs should include source address information to determine where the join attempts are originating from. In a scenario where an unknown script or process is the offending client, these log messages should have some information about the remote IP, or any other identifying information. Without that, these logs aren't actionable and can be excessively noisy.

### Current behavior:
The audit logs flood with "Instance Join Failed" messages without providing any helpful information such as the source address of the attempt. This makes it challenging to diagnose issues or identify the origin of failed join attempts.

This audit event was introduced in #40329 / [v15.2.4](https://github.com/gravitational/teleport/tree/v15.2.4)

### Bug details:
- **Teleport version**: 15.3.0 (Original report was from a Teleport Cloud user, but self-hosted is affected as well)
- **Recreation steps**:
  1. Observe an "Instance Join Failed" message in the audit log.
  2. Notice the absence of source address information in the log entry.
- **Debug logs**:
```json
{
  "event": "instance.join",
  "success": false,
  "time": "2024-05-10T16:05:44.933Z",
  "cluster_name": "example.teleort.sh",
  "token_expires": "0001-01-01T00:00:00Z",
  "uid": "fcd67e48-fc3e-4faf-9dbf-1a5b7edb299c",
  "error": "rpc error: code = Canceled desc = context canceled",
  "code": "TJ002E",
  "ei": 0
}
```

The lack of source address information in "Instance Join Failed" messages impedes troubleshooting efforts for Teleport administrators. It is valuable to include such details to facilitate better incident analysis and response.

### PR DESCRIPTION
**Pull request by strideynet**
Closes https://github.com/gravitational/teleport/issues/41421 partially - there's still a few potential cases where due to how we propagate the client IP, we actually don't know it until part of the way through the joining process (for challenge-response methods like IAM). We could eventually refactor to improve this and get this information at the start of the process.

changelog: Adds the remote address to audit log events emitted when a join for a Bot or Instance fails or succeeds.

### CODE REVIEW

File: lib/auth/join_azure.go [hunk 1]
```diff
@@ -372,6 +372,7 @@ func (a *Server) RegisterUsingAzureMethod(
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
+	joinRequest = req.RegisterUsingTokenRequest
```
Comment (hugoShaka) [inline lib/auth/join_azure.go:375]: So before we were always calling `handleJoinFailure` with `nil`?

Also if I understand correctly, we don't log any information server-side if the challenge failes?

Comment (hugoShaka) [reply lib/auth/join_azure.go:375]: Should this line be before the challenge block so we can log the details of the failed challenges?

Comment (strideynet) [reply lib/auth/join_azure.go:375]: Unfortunately previous design has shot us in the foot here - we don't know the RegisterUsingTokenRequest before we've issued the challenge and received the response from the client. I have no clue why the client doesn't send this until we've issued the challenge /shrug

Comment (hugoShaka) [reply lib/auth/join_azure.go:375]: 😭 This makes troubleshooting failed joins way harder

**Review by hugoShaka — APPROVED** [review]

**Comment by strideynet** [comment]
@justinas are you able to take a look at this tomorrow, I'd like to try and get this shipped in the next release to unblock a customer's troubleshooting on Cloud.

**Review by marcoandredinis — APPROVED** [review]

**Pull request merged**